src/backend/gl/src/window/surfman.rs

```diff
@@ -0,0 +1,346 @@
+//! [Surfman](https://github.com/pcwalton/surfman)-based OpenGL backend for GFX-hal
+
+use crate::{conv, native, Backend as B, Device, GlContainer, PhysicalDevice, QueueFamily, Starc};
+use hal::{adapter::Adapter, format as f, image, window};
+
+use arrayvec::ArrayVec;
+use glow::HasContext;
+use parking_lot::RwLock;
+use surfman as sm;
+
+use std::cell::RefCell;
+use std::fmt;
+use std::iter;
+
+#[derive(Debug)]
+pub struct Swapchain {
+    // Underlying window, required for presentation
+    pub(crate) context: Starc<RwLock<sm::Context>>,
+    // Extent because the window lies
+    pub(crate) extent: window::Extent2D,
+    ///
+    pub(crate) fbos: ArrayVec<[native::RawFrameBuffer; 3]>,
+}
+
+impl window::Swapchain<B> for Swapchain {
+    unsafe fn acquire_image(
+        &mut self,
+        _timeout_ns: u64,
+        _semaphore: Option<&native::Semaphore>,
+        _fence: Option<&native::Fence>,
+    ) -> Result<(window::SwapImageIndex, Option<window::Suboptimal>), window::AcquireError> {
+        // TODO: sync
+        Ok((0, None))
+    }
+}
+
+thread_local! {
+    /// The thread-local surfman connection
+    static SM_CONN: RefCell<sm::Connection> =
+        RefCell::new(sm::Connection::new().expect("TODO"));
+}
+
+pub struct Instance {
+    hardware_adapter: sm::Adapter,
+    // TODO: We're not using these yet, but leave them here for later
+    #[allow(dead_code)]
+    low_power_adapter: sm::Adapter,
+    #[allow(dead_code)]
+    software_adapter: sm::Adapter,
+}
+
+impl fmt::Debug for Instance {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Instance").field(&["Adapter..."; 3]).finish()
+    }
+}
+
+impl Instance {
+    fn get_default_context_attributes() -> sm::ContextAttributes {
+        sm::ContextAttributes {
+            version: sm::GLVersion::new(3, 3), // TODO: Figure out how to determine GL version
```

Comment:
is that for surfman to create a context internally? One reasonable default is:
  - OpenGL ES 3.0 on mobile
  - OpenGL Core 3.2 on macOS
  - OpenGL 2.1 elsewhere

Reply:
> is that for surfman to create a context internally? 

I think so. Its uses those attributes when creating GL contexts.

I'll add those defaults you mentioned.

Reply:
I tested out using version 2.1 on Linux, but it panics with a shader compilation error:

    thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: CompilationFailed("0:6(16): error: interface blocks with an instance name are not allowed in GLSL 1.40 (GLSL 1.50 or GLSL ES 3.00 required)\n")

It won't work until I go all the way up to 3.3.

Reply:
looks like SPIRV-Cross needs to be configured properly. It's fine, we can do it later :)